Feat/ticks taker tv

lnist opened the pull request:
This is the taker-side of " Internally implement everything in terms of tick/volume, then add interfaces in gives/wants" from https://github.com/mangrovedao/mangrove-core/pull/380

Comments have in general not been updated.

Commits are hopefully self-explanatory in their title, but some comments:
1 `feat(taker): Change average limit order to limit order and apply tick further out` - Changed to limit order by using tick, note we currently do not have a test for difference between average limit and limit. Also, a snipe-test was changed in a hacky way, should be updated by changing snipe to a tick-based API.
2. `feat(fillVolume): Use fillVolume instead of calculating gives/wants` - here "sor.gives" is renamed to "sor.givesToThisOffer" - this can be changed back but is useful to see changes.
3. `feat(stack): Reduce stack for consumption for market order` - introduces a partialFill for execute - similar result could be achieved by packing gives/wants into a single uint. It leaves a TODO for further improvement.
4. `fix(marketOrder): Do not store initialWants/gives` - sor.givesToThisOffer can be used instead.
5. `feat(marketOrder): Introduce tick-based API and overloads`: Introduction of new APIs

This is still WIP and the last commit opens up the discussion on what the API should be and what should be in libs/helpers and what should be the raw API.

Other commits can also be removed if we decide so.

lnist committed:
feat(taker): Change average limit order to limit order and apply tick further out

lnist committed:
feat(fillVolume): Use fillVolume instead of calculating gives/wants

lnist committed:
feat(stack): Reduce stack for consumption for market order

lnist committed:
fix(marketOrder): Do not store initialWants/gives

lnist committed:
feat(marketOrder): Introduce tick-based API and overloads

adhusson commented on lib/TickLib.sol:
```diff
@@ -178,6 +178,14 @@ library TickLib {
     }
   }
 
+  function tickFromTakerVolumes(uint takerGives, uint takerWants) internal pure returns (Tick) {
```
With average limit order, it is possible to have 

- wants=N, gives=MAX, fillWants=true (means "buy N at any price")

as well as : 

- wants=0, gives=N, fillWants=false (means "sell N at any price")

I think the current PR reproduces that, but maybe we should add tests to check it?

adhusson replied:
it seems ok

adhusson commented on src/MgvOfferTaking.sol:
```diff
@@ -35,11 +36,30 @@ abstract contract MgvOfferTaking is MgvHasOffers {
+    return marketOrder(outbound_tkn, inbound_tkn, maxTick, fillVolume, fillWants);
+  }
+
+  function marketOrder(address outbound_tkn, address inbound_tkn, int limitTick, uint fillVolume, bool fillWants)
```
needs tests for when tick/fillVolume are outside range

adhusson replied:
done

adhusson commented on src/MgvOfferTaking.sol:
```diff
@@ -35,11 +36,30 @@ abstract contract MgvOfferTaking is MgvHasOffers {
+    return marketOrder(outbound_tkn, inbound_tkn, maxTick, fillVolume, fillWants);
+  }
+
+  function marketOrderPrice(address outbound_tkn, address inbound_tkn, uint priceLimit, uint fillVolume, bool fillWants)
```
needs tests for when price/fillVolume are outside range

adhusson commented on src/MgvOfferTaking.sol:
```diff
@@ -191,24 +209,22 @@ abstract contract MgvOfferTaking is MgvHasOffers {
         /* Post-execution, `sor.wants`/`sor.gives` reflect how much was sent/taken by the offer. We will need it after the recursive call, so we save it in local variables. Same goes for `offerId`, `sor.offer` and `sor.offerDetail`. */
 
-        (gasused, makerData, mgvData) = execute(pair, mor, sor);
+        bool partialFill;
```
Needs more doc on why this variable is required now

adhusson replied:
variable removed for now

adhusson commented on src/MgvOfferTaking.sol:
```diff
@@ -191,24 +209,22 @@ abstract contract MgvOfferTaking is MgvHasOffers {
         /* If an execution was attempted, we move `sor` to the next offer. Note that the current state is inconsistent, since we have not yet updated `sor.offerDetails`. */
         if (mgvData != "mgv/notExecuted") {
-          sor.wants = mor.initialWants > mor.totalGot ? mor.initialWants - mor.totalGot : 0;
+          mor.fillVolume -= mor.fillWants ? sor.wantsFromThisOffer : sor.givesToThisOffer;
```
do we know that there will not be an underflow?

adhusson replied:
replying to myself: yes we do! added comments to that effect in `execute`

adhusson committed:
fix: correctly update `fillVolume`

since `initialWants` and `initialGives` have been removed, `fillVolume =
initialWants - totalGot` (for instance) is no longer possible and was replaced by `fillVolume = fillVolume - sor.wants`.

Since sor.wants does not depend on whether offer execution was
successful, we now update `fillVolume` only after a successful
execution.

Also added a regression test in TakerOperation.

adhusson committed:
fix: correct posthook values for sor.wants and sor.gives

The previously introduced optimization not restoring `sor.wants`/`gives`
values before a posthook when an offer has been partially filled does
not work when the partial fill has failed.

In that case, execution continues (despite the existence of a partial
fill), and so the partial-fill specific values should be restored
anyway.

Added a regression test in MakerPosthook.

adhusson committed:
cosmetic: return to old sor.wants/gives names

adhusson committed:
mark failing test to fix later

adhusson committed:
create 3 versions of marketOrderBy(Volume,Price,Tick)

adhusson committed:
add extrema checking + polish notation for maxPrice

adhusson committed:
better extrema checing for ticks

lnist committed:
Merge branch 'feat/ticks' into feat/ticks-taker-tv
